### DISCUSSION TITLE
How to get router.query value in  useEffect hook.

### DISCUSSION THREAD

**Discussion by ashx404** [vercel/next.js#12661] [Help]
I need to call an api on the basis of some query values but I am only able to get params when I use router.query within useEffect and not getting any query params.

**Comment by brianlovin** [comment]
Are you doing static generation? If so `query` will always be empty at build time. See https://nextjs.org/docs/api-reference/next/router#router-object

**Reply by ashx404** [reply]
Can I get query in a normal page? if I create a page and do not use getStaticProps and getStaticPaths?

**Reply by brianlovin** [reply]
Yes, the query will be available when the client renders and rehydrates:

```js
import { useRouter } from 'next/router'

export default function Component() {
  const router = useRouter()
  const { query } = router

  // do things
}
```

**Reply by ashx404** [reply]
Hey I changed to a class component because a functional component was not working, now I am trying to use withRouter, it asks for getInitialProps.
Also even in the functional component query was not accessible in the first render, even I wasn't able to access params.

**Reply by brianlovin** [reply]
If you can share some sample code, that would be great. `getInitialProps` has been deprecated so I'm not quite sure what the message is about :)

**Comment by IRediTOTO** [comment]
have same problem and here is my solution
```
const [initialValue,setInitialValue]=useState(null)
   useEffect(()=>{
      if(props.initialValue!=null){
         setInitialValue(props.initialValue)
      }
   },[props.initialValue])
   return (
      <Editor
         initialValue={initialValue?initialValue:""} />
```

**Comment by ashx404** [comment]
There is still no workaround on this? 
I am not able to understand the reason, I am using useRouter inside my component and query results in an empty object but asPath contains it in URL encoded form.

**Reply by benwinding** [reply]
This worked well, I found around from [this comment][1]

**Add `useQuery.ts` helper file**

``` js
// useQuery.js
import { useRouter } from 'next/router';

// Resolves query or returns null
export default function useQuery() {
  const router = useRouter();
  const hasQueryParams =
    /\[.+\]/.test(router.route) || /\?./.test(router.asPath);
  const ready = !hasQueryParams || Object.keys(router.query).length > 0;
  if (!ready) return null;
  return router.query;
}
```
**usage**

``` js
// In your components (instead of useRouter)
const query = useQuery();

useEffect(() => {
  if (!query) {
    return;
  }
  console.log('my query exists!!', query);
}, [query]);
```

  [1]: https://github.com/vercel/next.js/issues/8259#issuecomment-650225962

**Comment by cesarkohl** [comment]
I've found a solution. Surely, you can do it using Next.js' getStaticProps and getStaticPaths but it's a good practice to keep things closely tied to the JS core.

````
useEffect(() => {
    const last = window.location.href.split('/').pop();
}
````

If you want to get the _last element_ outside the useEffect() call the useRouter():

````
const router = useRouter();
const { last } = router.query;
````

Since my project doesn't need additional pages being generated at build time this is enough. If this is not your case then it's probably better to use getStaticProps and getStaticPaths.

**Reply by esiebomaj** [reply]
for some reason this does not work for me

**Comment by doxylee** [comment]
I found something:
> isReady: boolean - Whether the router fields are updated client-side and ready for use. Should only be used inside of useEffect methods and not for conditionally rendering on the server.
> https://nextjs.org/docs/api-reference/next/router#router-object

And the code would be like:
``` typescript
const router = useRouter();
useEffect(()=>{
    if(!router.isReady) return;

    // codes using router.query

}, [router.isReady]);
```

**Reply by bonbisu** [reply]
Best solution! Thanks!

**Reply by Naughty1905** [reply]
You saved my time! Thanks!

**Reply by ghost** [reply]
This is it!!! Thank you!!!! Saved me from chucking my laptop hehe

**Reply by ArnauldAnguh** [reply]
I almost went nuts. I used so many workarounds but didn't succeed. this just saved me 12 hrs of time waisted on trying to update a book record on a web app am working on

**Reply by Tijsvl** [reply]
Signed in just to say: Thank you!!!

**Reply by sami616** [reply]
Thanks a lot. this is the one!

**Reply by inspireSen** [reply]
Life saver! Thanks!

**Reply by MilesSRC** [reply]
Thank you! Saved me from le unforgiving spaghetti code!

**Reply by dawsnap** [reply]
Great solution! Thanks

**Reply by mihir0699** [reply]
Best solution ever !

**Reply by molikkidd** [reply]
Saved me a headache! Thank you

**Reply by dryleaf** [reply]
Well done

**Reply by 18-28** [reply]
so simple, thank you

**Reply by BennyG93** [reply]
Seems this thread of thank yous is very active... I'll add one more, thank you!

**Reply by abhijith-dev** [reply]
thank you !!!

**Reply by imbasri** [reply]
thanks for curing my headache 😊

**Reply by gitlango** [reply]
My Hero!

**Reply by FahimHassanMollah** [reply]
Thanks a lot.

**Reply by CarlosOV** [reply]
Thank you!!!

**Reply by lamlam-l** [reply]
lol its work like a champ

**Reply by SohailSayed** [reply]
Is there a good solution like this on NextJS 13? isReady got deprecated :(

**Reply by usotsuke** [reply]
U ARE THE BEST!!!!!

**Comment by njungenjenga** [comment]
If you're yet to upgrade to v10.0.5 > this worked for me:
```js  
const router = useRouter();
  useEffect(() => {
    if (Object.keys(router.query).length > 0) {
        // code using router.query
    }
  }, [router.query]);
  ```

**Reply by robertseghedi** [reply]
> If you're yet to upgrade to v10.0.5 > this worked for me:
> 
> ```js
> const router = useRouter();
>   useEffect(() => {
>     if (Object.keys(router.query).length > 0) {
>         // code using router.query
>     }
>   }, [router.query]);
> ```

Worked for me too.

**Comment by SohailSayed** [comment]
Anyone got a working solution for NextJS 13??

**Reply by BennyG93** [reply]
I think you could use the example in the official [docs](https://nextjs.org/docs/app/api-reference/functions/use-router#router-events).

According to the docs: The query object has been removed and is replaced by [useSearchParams()](https://nextjs.org/docs/app/api-reference/functions/use-search-params)

```
import { useEffect } from 'react'
import { usePathname, useSearchParams } from 'next/navigation'
[...]

  const pathname = usePathname()
  const searchParams = useSearchParams()

  useEffect(() => {
    const url = `${pathname}?${searchParams}`
    console.log(url)
    // You can now use the current URL
    // ...
  }, [pathname, searchParams])

[...]
```
